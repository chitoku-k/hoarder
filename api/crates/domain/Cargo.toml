[package]
name = "domain"
authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
doctest = false

[dependencies.chrono]
version = "0.4.39"
features = ["clock", "std"]
default-features = false

[dependencies.derive_more]
version = "2.0.1"
features = ["constructor", "deref", "display", "from"]

[dependencies.futures]
version = "0.3.31"
features = ["std"]
default-features = false

[dependencies.indenter]
version = "0.3.3"

[dependencies.ordermap]
version = "0.5.5"

[dependencies.percent-encoding]
version = "2.3.1"

[dependencies.regex]
version = "1.11.1"

[dependencies.serde]
version = "1.0.218"
features = ["derive"]

[dependencies.strum]
version = "0.27.0"
features = ["derive"]

[dependencies.thiserror]
version = "2.0.11"

[dependencies.tokio]
version = "1.43.0"
features = ["macros", "rt-multi-thread"]

[dependencies.tracing]
version = "0.1.41"

[dependencies.uuid]
version = "1.13.1"
features = ["serde", "v4"]

[dev-dependencies.anyhow]
version = "1.0.95"

[dev-dependencies.indoc]
version = "2.0.5"

[dev-dependencies.mockall]
version = "0.13.1"

[dev-dependencies.pretty_assertions]
version = "1.4.1"
features = ["unstable"]

[dev-dependencies.serial_test]
version = "3.2.0"
