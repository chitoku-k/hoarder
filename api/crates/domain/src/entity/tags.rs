use std::collections::BTreeSet;

use chrono::{DateTime, Utc};
use derive_more::{Constructor, Deref, Display, From};
use serde::Serialize;
use uuid::Uuid;

#[derive(Clone, Copy, Debug, Default, Deref, Display, Eq, From, Hash, Ord, PartialEq, PartialOrd)]
pub struct TagId(Uuid);

#[derive(Clone, Constructor, Copy, Debug, Eq, From, PartialEq)]
pub struct TagDepth {
    parent: u32,
    children: u32,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct Tag {
    pub id: TagId,
    pub name: String,
    pub kana: String,
    pub aliases: AliasSet,
    pub parent: Option<Box<Self>>,
    pub children: Vec<Self>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl TagId {
    pub const fn root() -> Self {
        Self(Uuid::nil())
    }

    pub const fn is_root(&self) -> bool {
        self.0.is_nil()
    }
}

impl TagDepth {
    pub const fn parent(&self) -> u32 {
        self.parent
    }

    pub const fn children(&self) -> u32 {
        self.children
    }

    pub const fn has_parent(&self) -> bool {
        self.parent > 0
    }

    pub const fn has_children(&self) -> bool {
        self.children > 0
    }
}

#[derive(Clone, Constructor, Debug, Default, Deref, Eq, PartialEq, Serialize)]
pub struct AliasSet(BTreeSet<String>);

impl AliasSet {
    pub fn add_all<T>(&mut self, aliases: T)
    where
        T: IntoIterator<Item = String>,
    {
        for alias in aliases {
            self.0.insert(alias);
        }
    }

    pub fn remove_all<T>(&mut self, aliases: T)
    where
        T: IntoIterator<Item = String>,
    {
        for alias in aliases {
            self.0.remove(&alias);
        }
    }

    pub fn into_inner(self) -> BTreeSet<String> {
        self.0
    }
}

impl From<Vec<String>> for AliasSet {
    fn from(v: Vec<String>) -> Self {
        Self::new(v.into_iter().collect())
    }
}

impl From<AliasSet> for Vec<String> {
    fn from(v: AliasSet) -> Self {
        v.0.into_iter().collect()
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn tag_id_root() {
        let actual = TagId::root();

        assert!(actual.0.is_nil());
    }

    #[test]
    fn tag_id_is_root() {
        let actual = TagId::root();

        assert!(actual.is_root());
    }

    #[test]
    fn tag_depth_parent() {
        let depth = TagDepth::new(1, 2);
        let actual = depth.parent();

        assert_eq!(actual, 1);
    }

    #[test]
    fn tag_depth_children() {
        let depth = TagDepth::new(1, 2);
        let actual = depth.children();

        assert_eq!(actual, 2);
    }

    #[test]
    fn tag_depth_has_parent() {
        let depth = TagDepth::new(1, 2);
        let actual = depth.has_parent();

        assert!(actual);
    }

    #[test]
    fn tag_depth_has_no_parent() {
        let depth = TagDepth::new(0, 2);
        let actual = depth.has_parent();

        assert!(!actual);
    }

    #[test]
    fn tag_depth_has_children() {
        let depth = TagDepth::new(1, 2);
        let actual = depth.has_children();

        assert!(actual);
    }

    #[test]
    fn tag_depth_has_no_children() {
        let depth = TagDepth::new(1, 0);
        let actual = depth.has_children();

        assert!(!actual);
    }

    #[test]
    fn alias_set_add_all() {
        let mut actual = AliasSet::new(BTreeSet::from(["baz".to_string()]));
        actual.add_all(["foo".to_string(), "bar".to_string()]);

        assert_eq!(actual, AliasSet::new(BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()])));
    }

    #[test]
    fn alias_set_remove_all() {
        let mut actual = AliasSet::new(BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()]));
        actual.remove_all(["foo".to_string(), "bar".to_string()]);

        assert_eq!(actual, AliasSet::new(BTreeSet::from(["baz".to_string()])));
    }

    #[test]
    fn alias_set_into_inner() {
        let set = AliasSet::new(BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()]));
        let actual = set.into_inner();

        assert_eq!(actual, BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()]));
    }

    #[test]
    fn alias_set_from_strings() {
        let actual = AliasSet::from(vec!["foo".to_string(), "bar".to_string(), "baz".to_string()]);

        assert_eq!(actual, AliasSet::new(BTreeSet::from(["foo".to_string(), "bar".to_string(), "baz".to_string()])));
    }
}
