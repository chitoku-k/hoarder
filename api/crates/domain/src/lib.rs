#![allow(clippy::too_many_arguments)]

pub mod entity;
pub mod parser;
pub mod processor;
pub mod repository;
pub mod service;
