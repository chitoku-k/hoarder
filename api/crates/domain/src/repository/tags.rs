use async_trait::async_trait;

use crate::{
    entity::tags::{Tag, TagDepth, TagId},
    repository::{DeleteResult, OrderDirection},
};

#[cfg_attr(test, mockall::automock)]
#[async_trait]
pub trait TagsRepository: Send + Sync + 'static {
    /// Creates a tag.
    async fn create(&self, name: &str, kana: &str, aliases: &[String], parent_id: Option<TagId>, depth: TagDepth) -> anyhow::Result<Tag>;

    /// Fetches tags by their IDs.
    async fn fetch_by_ids<T>(&self, ids: T, depth: TagDepth) -> anyhow::Result<Vec<Tag>>
    where
        T: IntoIterator<Item = TagId> + Send + Sync + 'static;

    /// Fetches tags by their names like the given parameter.
    async fn fetch_by_name_or_alias_like(&self, name_or_alias_like: &str, depth: TagDepth) -> anyhow::Result<Vec<Tag>>;

    /// Fetches all tags.
    async fn fetch_all(&self, depth: TagDepth, root: bool, after: Option<(String, TagId)>, before: Option<(String, TagId)>, order: OrderDirection, limit: u64) -> anyhow::Result<Vec<Tag>>;

    /// Updates the tag by ID.
    async fn update_by_id<T, U>(
        &self,
        id: TagId,
        name: Option<String>,
        kana: Option<String>,
        add_aliases: T,
        remove_aliases: U,
        depth: TagDepth,
    ) -> anyhow::Result<Tag>
    where
        T: IntoIterator<Item = String> + Send + Sync + 'static,
        U: IntoIterator<Item = String> + Send + Sync + 'static;

    /// Attaches the tag to the existing tag by ID.
    async fn attach_by_id(&self, id: TagId, parent_id: TagId, depth: TagDepth) -> anyhow::Result<Tag>;

    /// Detaches the tag from its parent by ID.
    async fn detach_by_id(&self, id: TagId, depth: TagDepth) -> anyhow::Result<Tag>;

    /// Deletes the tag by ID.
    async fn delete_by_id(&self, id: TagId, recursive: bool) -> anyhow::Result<DeleteResult>;
}
