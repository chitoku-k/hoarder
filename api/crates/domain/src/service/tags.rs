use async_trait::async_trait;
use derive_more::Constructor;

use crate::{
    entity::{
        tag_types::{TagType, TagTypeId},
        tags::{Tag, TagDepth, TagId},
    },
    repository::{tag_types, tags, DeleteResult, Direction, Order},
};

#[cfg_attr(feature = "test-mock", mockall::automock)]
#[async_trait]
pub trait TagsServiceInterface: Send + Sync + 'static {
    /// Creates a tag.
    async fn create_tag(&self, name: &str, kana: &str, aliases: &[String], parent_id: Option<TagId>, depth: TagDepth) -> anyhow::Result<Tag>;

    /// Creates a tag type.
    async fn create_tag_type(&self, slug: &str, name: &str) -> anyhow::Result<TagType>;

    /// Gets tags.
    async fn get_tags(
        &self,
        depth: TagDepth,
        root: bool,
        cursor: Option<(String, TagId)>,
        order: Order,
        direction: Direction,
        limit: u64,
    ) -> anyhow::Result<Vec<Tag>>;

    /// Gets the tags by their IDs.
    async fn get_tags_by_ids<T>(&self, ids: T, depth: TagDepth) -> anyhow::Result<Vec<Tag>>
    where
        T: IntoIterator<Item = TagId> + Send + Sync + 'static;

    /// Gets the tags by their name or alias.
    async fn get_tags_by_name_or_alias_like(&self, name_or_alias_like: &str, depth: TagDepth) -> anyhow::Result<Vec<Tag>>;

    /// Gets tag types.
    async fn get_tag_types(&self) -> anyhow::Result<Vec<TagType>>;

    /// Updates the tag by ID.
    async fn update_tag_by_id<T, U>(&self, id: TagId, name: Option<String>, kana: Option<String>, add_aliases: T, remove_aliases: U, depth: TagDepth) -> anyhow::Result<Tag>
    where
        T: IntoIterator<Item = String> + Send + Sync + 'static,
        U: IntoIterator<Item = String> + Send + Sync + 'static;

    /// Updates the tag type by ID.
    async fn update_tag_type_by_id<'a, 'b>(&self, id: TagTypeId, slug: Option<&'a str>, name: Option<&'b str>) -> anyhow::Result<TagType>;

    /// Attaches the tag by ID.
    async fn attach_tag_by_id(&self, id: TagId, parent_id: TagId, depth: TagDepth) -> anyhow::Result<Tag>;

    /// Detaches the tag by ID.
    async fn detach_tag_by_id(&self, id: TagId, depth: TagDepth) -> anyhow::Result<Tag>;

    /// Delete the tag by ID.
    async fn delete_tag_by_id(&self, id: TagId, recursive: bool) -> anyhow::Result<DeleteResult>;

    /// Delete the tag type by ID.
    async fn delete_tag_type_by_id(&self, id: TagTypeId) -> anyhow::Result<DeleteResult>;
}

#[derive(Clone, Constructor)]
pub struct TagsService<TagsRepository, TagTypesRepository> {
    tags_repository: TagsRepository,
    tag_types_repository: TagTypesRepository,
}

#[async_trait]
impl<TagsRepository, TagTypesRepository> TagsServiceInterface for TagsService<TagsRepository, TagTypesRepository>
where
    TagsRepository: tags::TagsRepository,
    TagTypesRepository: tag_types::TagTypesRepository,
{
    async fn create_tag(&self, name: &str, kana: &str, aliases: &[String], parent_id: Option<TagId>, depth: TagDepth) -> anyhow::Result<Tag> {
        match self.tags_repository.create(name, kana, aliases, parent_id, depth).await {
            Ok(tag) => Ok(tag),
            Err(e) => {
                log::error!("failed to create a tag\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn create_tag_type(&self, slug: &str, name: &str) -> anyhow::Result<TagType> {
        match self.tag_types_repository.create(slug, name).await {
            Ok(tag_type) => Ok(tag_type),
            Err(e) => {
                log::error!("failed to create a tag type\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn get_tags(
        &self,
        depth: TagDepth,
        root: bool,
        cursor: Option<(String, TagId)>,
        order: Order,
        direction: Direction,
        limit: u64,
    ) -> anyhow::Result<Vec<Tag>> {
        match self.tags_repository.fetch_all(depth, root, cursor, order, direction, limit).await {
            Ok(tags) => Ok(tags),
            Err(e) => {
                log::error!("failed to get tags\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn get_tags_by_ids<T>(&self, ids: T, depth: TagDepth) -> anyhow::Result<Vec<Tag>>
    where
        T: IntoIterator<Item = TagId> + Send + Sync + 'static,
    {
        match self.tags_repository.fetch_by_ids(ids, depth).await {
            Ok(tags) => Ok(tags),
            Err(e) => {
                log::error!("failed to get tags\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn get_tags_by_name_or_alias_like(&self, name_or_alias_like: &str, depth: TagDepth) -> anyhow::Result<Vec<Tag>> {
        match self.tags_repository.fetch_by_name_or_alias_like(name_or_alias_like, depth).await {
            Ok(tags) => Ok(tags),
            Err(e) => {
                log::error!("failed to get tags\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn get_tag_types(&self) -> anyhow::Result<Vec<TagType>> {
        match self.tag_types_repository.fetch_all().await {
            Ok(tag_types) => Ok(tag_types),
            Err(e) => {
                log::error!("failed to get tag types\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn update_tag_by_id<T, U>(&self, id: TagId, name: Option<String>, kana: Option<String>, add_aliases: T, remove_aliases: U, depth: TagDepth) -> anyhow::Result<Tag>
    where
        T: IntoIterator<Item = String> + Send + Sync + 'static,
        U: IntoIterator<Item = String> + Send + Sync + 'static,
    {
        match self.tags_repository.update_by_id(id, name, kana, add_aliases, remove_aliases, depth).await {
            Ok(tag) => Ok(tag),
            Err(e) => {
                log::error!("failed to update the tag\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn update_tag_type_by_id<'a, 'b>(&self, id: TagTypeId, slug: Option<&'a str>, name: Option<&'b str>) -> anyhow::Result<TagType> {
        match self.tag_types_repository.update_by_id(id, slug, name).await {
            Ok(tag_type) => Ok(tag_type),
            Err(e) => {
                log::error!("failed to update the tag type\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn attach_tag_by_id(&self, id: TagId, parent_id: TagId, depth: TagDepth) -> anyhow::Result<Tag> {
        match self.tags_repository.attach_by_id(id, parent_id, depth).await {
            Ok(tag) => Ok(tag),
            Err(e) => {
                log::error!("failed to attach the tag\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn detach_tag_by_id(&self, id: TagId, depth: TagDepth) -> anyhow::Result<Tag> {
        match self.tags_repository.detach_by_id(id, depth).await {
            Ok(tag) => Ok(tag),
            Err(e) => {
                log::error!("failed to detach the tag\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn delete_tag_by_id(&self, id: TagId, recursive: bool) -> anyhow::Result<DeleteResult> {
        match self.tags_repository.delete_by_id(id, recursive).await {
            Ok(result) => Ok(result),
            Err(e) => {
                log::error!("failed to delete the tag\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn delete_tag_type_by_id(&self, id: TagTypeId) -> anyhow::Result<DeleteResult> {
        match self.tag_types_repository.delete_by_id(id).await {
            Ok(result) => Ok(result),
            Err(e) => {
                log::error!("failed to delete the tag type\nError: {e:?}");
                Err(e)
            },
        }
    }
}
