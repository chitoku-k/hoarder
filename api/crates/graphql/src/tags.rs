use std::collections::BTreeSet;

use async_graphql::{
    connection::{CursorType, DefaultEdgeName, Edge, EmptyFields},
    InputObject, Lookahead, SimpleObject,
};
use base64::{prelude::BASE64_STANDARD, Engine};
use chrono::{DateTime, Utc};
use derive_more::Constructor;
use domain::entity::{
    tag_types::{self, TagTypeId},
    tags::{self, TagDepth, TagId},
};
use uuid::Uuid;

use crate::error::ErrorKind;

/// A tag represents a user-friendly and hierarchical attribute attached to media.
#[derive(SimpleObject)]
pub struct Tag {
    /// The ID of the Tag object.
    id: Uuid,
    /// The name of the tag.
    name: String,
    /// The kana of the tag.
    kana: String,
    /// The list of aliases for the tag.
    aliases: BTreeSet<String>,
    /// The parent node of the tag.
    parent: Option<Box<Tag>>,
    /// The child nodes of the tag.
    children: Vec<Tag>,
    /// The date at which the tag was created.
    created_at: DateTime<Utc>,
    /// The date at which the tag was updated.
    updated_at: DateTime<Utc>,
}

#[derive(Clone, Constructor, Debug, Eq, PartialEq)]
pub(crate) struct TagCursor(String, Uuid);

/// A tag type represents a type of the tag being attached to media describing how
/// the tag corresponds to the media.
#[derive(SimpleObject)]
pub struct TagType {
    /// The ID of the TagType object.
    id: Uuid,
    /// The short and user-friendly name that uniquely identifies the tag type.
    slug: String,
    /// The name of the tag type.
    name: String,
    /// The kana of the tag type.
    kana: String,
}

/// A tag tag type is a pair of a tag and a tag type.
#[derive(Constructor, SimpleObject)]
pub struct TagTagType {
    /// The tag.
    tag: Tag,
    /// The tag type.
    #[graphql(name = "type")]
    tag_type: TagType,
}

/// A tag tag type input is a pair of the ID of a tag and tag type.
#[derive(Clone, Copy, InputObject)]
pub struct TagTagTypeInput {
    /// The ID of the tag.
    tag_id: Uuid,
    /// The ID of the tag type.
    tag_type_id: Uuid,
}

impl From<tags::Tag> for Tag {
    fn from(tag: tags::Tag) -> Self {
        let parent = tag.parent
            .map(|p| (*p).into())
            .map(Box::new);

        let children = tag.children
            .into_iter()
            .map(Into::into)
            .collect();

        Self {
            id: *tag.id,
            name: tag.name,
            kana: tag.kana,
            aliases: tag.aliases.into_inner(),
            parent,
            children,
            created_at: tag.created_at,
            updated_at: tag.updated_at,
        }
    }
}

impl TagCursor {
    const DELIMITER: char = '\x00';

    pub fn into_inner(self) -> (String, TagId) {
        (self.0, self.1.into())
    }
}

impl CursorType for TagCursor {
    type Error = ErrorKind;

    fn decode_cursor(s: &str) -> Result<Self, Self::Error> {
        let bin = BASE64_STANDARD.decode(s).map_err(|_| ErrorKind::CursorInvalid)?;
        let str = String::from_utf8(bin).map_err(|_| ErrorKind::CursorInvalid)?;

        let (kana, uuid) = str.split_once(Self::DELIMITER).ok_or(ErrorKind::CursorInvalid)?;
        let kana = kana.to_string();
        let uuid = Uuid::parse_str(uuid).map_err(|_| ErrorKind::CursorInvalid)?;

        Ok(TagCursor::new(kana, uuid))
    }

    fn encode_cursor(&self) -> String {
        let str = format!("{}{}{}", &self.0, Self::DELIMITER, &self.1);

        BASE64_STANDARD.encode(str)
    }
}

impl From<Tag> for Edge<TagCursor, Tag, EmptyFields, DefaultEdgeName> {
    fn from(tag: Tag) -> Self {
        let cursor = TagCursor::new(tag.kana.clone(), tag.id);
        Edge::new(cursor, tag)
    }
}

impl From<tag_types::TagType> for TagType {
    fn from(tag_type: tag_types::TagType) -> Self {
        Self {
            id: *tag_type.id,
            slug: tag_type.slug,
            name: tag_type.name,
            kana: tag_type.kana,
        }
    }
}

impl From<TagTagTypeInput> for (TagId, TagTypeId) {
    fn from(input: TagTagTypeInput) -> Self {
        (input.tag_id.into(), input.tag_type_id.into())
    }
}

pub fn get_tag_depth(root: &Lookahead<'_>) -> TagDepth {
    let mut parent = 0;
    let mut children = 0;

    let mut parent_look_ahead = root.field("parent");
    let mut children_look_ahead = root.field("children");

    loop {
        match (parent_look_ahead.exists(), children_look_ahead.exists()) {
            (false, false) => break,
            (true, true) => {
                parent += 1;
                children += 1;
            },
            (true, false) => {
                parent += 1;
            },
            (false, true) => {
                children += 1;
            },
        }

        parent_look_ahead = parent_look_ahead.field("parent");
        children_look_ahead = children_look_ahead.field("children");
    }

    TagDepth::new(parent, children)
}
