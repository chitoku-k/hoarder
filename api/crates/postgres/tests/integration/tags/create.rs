use std::collections::BTreeSet;

use domain::{
    entity::tags::{AliasSet, Tag, TagDepth, TagId},
    repository::tags::TagsRepository,
};
use chrono::{TimeZone, Utc};
use futures::TryStreamExt;
use postgres::tags::PostgresTagsRepository;
use pretty_assertions::assert_eq;
use sqlx::Row;
use test_context::test_context;
use uuid::{uuid, Uuid};

use super::DatabaseContext;

#[test_context(DatabaseContext)]
#[tokio::test]
async fn with_parent_succeeds(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.create(
        "七森中☆生徒会",
        "ななもりちゅうせいとかい",
        ["生徒会".to_string(), "七森中生徒会".to_string()].into_iter(),
        Some(TagId::from(uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"))),
        TagDepth::new(2, 2),
    ).await.unwrap();

    let actual_id = actual.id;
    assert_eq!(actual.name, "七森中☆生徒会".to_string());
    assert_eq!(actual.kana, "ななもりちゅうせいとかい");
    assert_eq!(actual.aliases, AliasSet::new(BTreeSet::from(["生徒会".to_string(), "七森中生徒会".to_string()])));
    assert_eq!(
        actual.parent,
        Some(Box::new(Tag {
            id: TagId::from(uuid!("744b7274-371b-4790-8f5a-df4d76e983ba")),
            name: "ゆるゆり".to_string(),
            kana: "ゆるゆり".to_string(),
            aliases: AliasSet::default(),
            parent: None,
            children: Vec::new(),
            created_at: Utc.with_ymd_and_hms(2022, 1, 2, 3, 4, 5).unwrap(),
            updated_at: Utc.with_ymd_and_hms(2022, 2, 3, 4, 5, 8).unwrap(),
        })),
    );
    assert_eq!(actual.children, Vec::new());

    let actual = sqlx::query(r#"SELECT "id", "name", "kana", "aliases" FROM "tags" WHERE "id" = $1"#)
        .bind(*actual.id)
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<&str, &str>("name"), "七森中☆生徒会");
    assert_eq!(actual.get::<&str, &str>("kana"), "ななもりちゅうせいとかい");
    assert_eq!(actual.get::<Vec<String>, &str>("aliases"), vec!["生徒会".to_string(), "七森中生徒会".to_string()]);

    let actual: Vec<_> = sqlx::query(r#"SELECT "ancestor_id", "descendant_id", "distance" FROM "tag_paths" WHERE "descendant_id" = $1 ORDER BY "distance" DESC"#)
        .bind(*actual_id)
        .fetch(&ctx.pool)
        .try_collect()
        .await
        .unwrap();

    assert_eq!(actual.len(), 3);

    assert_eq!(actual[0].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[0].get::<Uuid, &str>("descendant_id"), *actual_id);
    assert_eq!(actual[0].get::<i32, &str>("distance"), 2);

    assert_eq!(actual[1].get::<Uuid, &str>("ancestor_id"), uuid!("744b7274-371b-4790-8f5a-df4d76e983ba"));
    assert_eq!(actual[1].get::<Uuid, &str>("descendant_id"), *actual_id);
    assert_eq!(actual[1].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[2].get::<Uuid, &str>("ancestor_id"), *actual_id);
    assert_eq!(actual[2].get::<Uuid, &str>("descendant_id"), *actual_id);
    assert_eq!(actual[2].get::<i32, &str>("distance"), 0);
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn without_parent_succeeds(ctx: &DatabaseContext) {
    let repository = PostgresTagsRepository::new(ctx.pool.clone());
    let actual = repository.create(
        "七森中☆生徒会",
        "ななもりちゅうせいとかい",
        ["生徒会".to_string(), "七森中生徒会".to_string()].into_iter(),
        None,
        TagDepth::new(2, 2),
    ).await.unwrap();

    let actual_id = actual.id;
    assert_eq!(actual.name, "七森中☆生徒会".to_string());
    assert_eq!(actual.kana, "ななもりちゅうせいとかい");
    assert_eq!(actual.aliases, AliasSet::new(BTreeSet::from(["生徒会".to_string(), "七森中生徒会".to_string()])));
    assert_eq!(actual.parent, None);
    assert_eq!(actual.children, Vec::new());

    let actual = sqlx::query(r#"SELECT "id", "name", "kana", "aliases" FROM "tags" WHERE "id" = $1"#)
        .bind(*actual.id)
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<&str, &str>("name"), "七森中☆生徒会");
    assert_eq!(actual.get::<&str, &str>("kana"), "ななもりちゅうせいとかい");
    assert_eq!(actual.get::<Vec<String>, &str>("aliases"), vec!["生徒会".to_string(), "七森中生徒会".to_string()]);

    let actual: Vec<_> = sqlx::query(r#"SELECT "ancestor_id", "descendant_id", "distance" FROM "tag_paths" WHERE "descendant_id" = $1 ORDER BY "distance" DESC"#)
        .bind(*actual_id)
        .fetch(&ctx.pool)
        .try_collect()
        .await
        .unwrap();

    assert_eq!(actual.len(), 2);

    assert_eq!(actual[0].get::<Uuid, &str>("ancestor_id"), uuid!("00000000-0000-0000-0000-000000000000"));
    assert_eq!(actual[0].get::<Uuid, &str>("descendant_id"), *actual_id);
    assert_eq!(actual[0].get::<i32, &str>("distance"), 1);

    assert_eq!(actual[1].get::<Uuid, &str>("ancestor_id"), *actual_id);
    assert_eq!(actual[1].get::<Uuid, &str>("descendant_id"), *actual_id);
    assert_eq!(actual[1].get::<i32, &str>("distance"), 0);
}
