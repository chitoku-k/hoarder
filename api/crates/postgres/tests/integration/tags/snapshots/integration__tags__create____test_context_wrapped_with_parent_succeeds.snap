---
source: crates/postgres/tests/integration/tags/create.rs
expression: ctx.queries()
---
[[queries]]
sql = '''
INSERT INTO
  "tags" ("name", "kana", "aliases")
VALUES
  ($1, $2, $3)
RETURNING
  "id"'''
rows_affected = 1
rows_returned = 1

[[queries]]
sql = '''
INSERT INTO
  "tag_paths" ("ancestor_id", "descendant_id", "distance")
VALUES
  ($1, $2, $3)'''
rows_affected = 1
rows_returned = 0

[[queries]]
sql = '''
INSERT INTO
  "tag_paths" ("ancestor_id", "descendant_id", "distance")
SELECT
  "ancestor_id",
  $1,
  "distance" + $2
FROM
  "tag_paths"
WHERE
  "descendant_id" = $3'''
rows_affected = 2
rows_returned = 0

[[queries]]
sql = '''
INSERT INTO
  "tag_paths" ("ancestor_id", "descendant_id", "distance")
SELECT
  "tag_path_ancestors"."ancestor_id",
  "tag_path_descendants"."descendant_id",
  "tag_path_ancestors"."distance" + "tag_path_descendants"."distance"
FROM
  "tag_paths" AS "tag_path_ancestors"
INNER JOIN
  "tag_paths" AS "tag_path_descendants" ON "tag_path_ancestors"."descendant_id" = "tag_path_descendants"."ancestor_id"
WHERE
  "tag_path_ancestors"."descendant_id" = $1
  AND "tag_path_descendants"."ancestor_id" = $2
  AND "tag_path_ancestors"."ancestor_id" <> $3
  AND "tag_path_descendants"."descendant_id" <> $4'''
rows_affected = 0
rows_returned = 0

[[queries]]
sql = 'RELEASE SAVEPOINT _sqlx_savepoint_1'
rows_affected = 0
rows_returned = 0

[[queries]]
sql = '''
SELECT
  "tag_orders"."order",
  "distance",
  "tag_ancestors"."id" AS "ancestor_id",
  "tag_ancestors"."name" AS "ancestor_name",
  "tag_ancestors"."kana" AS "ancestor_kana",
  "tag_ancestors"."aliases" AS "ancestor_aliases",
  "tag_ancestors"."created_at" AS "ancestor_created_at",
  "tag_ancestors"."updated_at" AS "ancestor_updated_at",
  "tag_descendants"."id" AS "descendant_id",
  "tag_descendants"."name" AS "descendant_name",
  "tag_descendants"."kana" AS "descendant_kana",
  "tag_descendants"."aliases" AS "descendant_aliases",
  "tag_descendants"."created_at" AS "descendant_created_at",
  "tag_descendants"."updated_at" AS "descendant_updated_at"
FROM
  "tag_paths"
LEFT JOIN
  (
    SELECT
      "column1" AS "order",
      "column2" AS "id"
    FROM
      (
        VALUES
          ($1, $2), ($3, $4)) AS "tag_orders"
  ) AS "tag_orders" ON "tag_orders"."id" = "descendant_id"
  AND "ancestor_id" IN ("descendant_id", $5)
INNER JOIN
  "tags" AS "tag_ancestors" ON "tag_ancestors"."id" = "tag_paths"."ancestor_id"
INNER JOIN
  "tags" AS "tag_descendants" ON "tag_descendants"."id" = "tag_paths"."descendant_id"
WHERE
  (
    "distance" = $6
    AND "ancestor_id" IN ($7)
  )
  OR (
    "distance" <= $8
    AND "descendant_id" IN (
      SELECT
        "ancestor_id"
      FROM
        "tag_paths"
      WHERE
        "distance" <= $9
        AND "descendant_id" IN ($10)
    )
  )
  OR (
    "distance" <= $11
    AND "ancestor_id" IN (
      SELECT
        "descendant_id"
      FROM
        "tag_paths"
      WHERE
        "distance" <= $12
        AND "ancestor_id" IN ($13)
    )
  )
ORDER BY
  "distance" ASC,
  "tag_orders"."order" ASC,
  "tag_ancestors"."kana" ASC,
  "tag_descendants"."kana" ASC'''
rows_affected = 5
rows_returned = 5

[[queries]]
sql = 'COMMIT'
rows_affected = 0
rows_returned = 0
